rclpy/src/rclpy/events_executor/timers_manager.cpp

```diff
@@ -0,0 +1,335 @@
+    if (!on_debug_time_ && next_ready_ns) {
+      next_update_wait_.expires_from_now(std::chrono::nanoseconds(*next_ready_ns));
+      next_update_wait_.async_wait([this](const asio::error_code & ec) {
+          if (!ec) {
+            UpdateTimers();
+          } else if (ec != asio::error::operation_aborted) {
+            throw std::runtime_error("Error waiting for next timer: " + ec.message());
+          }
+      });
+    } else {
+      next_update_wait_.cancel();
+    }
+  }
+
+  /// Returns the number of nanoseconds until the next callback on the given timer is due.  Value
+  /// may be negative or zero if callback time has already been reached.  Returns std::nullopt if
+  /// the timer is canceled.
+  static std::optional<int64_t> GetNextCallNanoseconds(const rcl_timer_t * rcl_timer)
+  {
+    int64_t time_until_next_call{};
+    const rcl_ret_t ret = rcl_timer_get_time_until_next_call(rcl_timer, &time_until_next_call);
+    switch (ret) {
+      case RCL_RET_OK:
+        return time_until_next_call;
+      case RCL_RET_TIMER_CANCELED:
+        return {};
+      default:
+        throw std::runtime_error(
+          std::string("Failed to fetch timer ready time: ") + rcl_get_error_string().str);
+    }
+  }
+
+  asio::any_io_executor executor_;
+  rcl_clock_t * const clock_;
+  ClockJumpCallbackT jump_cb_;
+  TimerResetCallbackT reset_cb_;
+  bool on_debug_time_{};
+
+  std::unordered_map<rcl_timer_t *, std::function<void()>> timers_;
+  asio::steady_timer next_update_wait_{executor_};
+};
+
+RclTimersManager::RclTimersManager(const asio::any_io_executor & executor)
+: executor_(executor) {}
+
+RclTimersManager::~RclTimersManager() {}
+
+namespace
+{
+rcl_clock_t * GetTimerClock(rcl_timer_t * timer)
+{
+  rcl_clock_t * clock{};
+  if (RCL_RET_OK != rcl_timer_clock(timer, &clock)) {
+    throw std::runtime_error(
+      std::string("Failed to determine clock for timer: ") + rcl_get_error_string().str);
+  }
+  return clock;
+}
+}  // namespace
+
+void RclTimersManager::AddTimer(rcl_timer_t * timer, std::function<void()> ready_callback)
+{
+  // Figure out the clock this timer is using, make sure a manager exists for that clock, then
+  // forward the timer to that clock's manager.
+  rcl_clock_t * clock = GetTimerClock(timer);
+  auto it = clock_managers_.find(clock);
+  if (it == clock_managers_.end()) {
+    std::tie(it, std::ignore) = clock_managers_.insert(
+      std::make_pair(clock, std::make_unique<ClockManager>(executor_, clock)));
+  }
+  it->second->AddTimer(timer, ready_callback);
+}
+
+void RclTimersManager::RemoveTimer(rcl_timer_t * timer)
+{
+  const rcl_clock_t * clock = GetTimerClock(timer);
+  auto it = clock_managers_.find(clock);
+  if (it == clock_managers_.end()) {
+    throw py::key_error("Attempt to remove timer from unmanaged clock");
+  }
+  it->second->RemoveTimer(timer);
+  if (it->second->empty()) {
+    clock_managers_.erase(it);
```

Comment:
hm, isn't there a race here, that the asio callback still happens into the now deleted clock_manager ?

Reply:
I think the timer's own callback is safe, because it calls a user callback directly without capturing the clock manager itself.  However, it may be possible for `HandleJump()` and `UpdateTimers()` to get called with a deleted `this` pointer, if you remove the last timer from a clock immediately after a callback was posted, and before the callback is dispatched.  I'll make a note to look at this closer later.

Reply:
I think this is resolved now.